src/mame/drivers/dfs500.cpp

```diff
@@ -0,0 +1,1036 @@
+}
+
+void dfs500_state::WB0_w(offs_t offset, uint16_t data)
+{
+	m_maincpu_latch16 = data;
+	m_TOE = true;
+}
+
+uint16_t dfs500_state::RB0_r(offs_t offset)
+{
+	return m_effectcpu_latch16;
+	m_TOC = false;
+}
+
+uint8_t dfs500_state::RB1_r(offs_t offset)
+{
+	//"TEST, OPT2, OPT1, VD, T2, T1, TOE, TOC"
+	uint8_t value = 0;
+	// FIXME! Add other signals.
+	if (m_TOE) value |= 0x02;
+	if (m_TOC) value |= 0x01;
+	return value;
+}
+
+uint8_t dfs500_state::RB2_r(offs_t offset)
+{
+	uint8_t value = 0;
+	value |= ((ioport("DSW_S1")->read() & 0x0f) << 4); // (Editing Control Unit Select)
+	value |= (ioport("DSW_S2")->read() & 0x0f);        // (Freeze Timing)
+	// TODO:
+	// if (m_BVS) value |= 0x10;
+	return value;
+}
+
+uint8_t dfs500_state::cpanel_reg0_r(offs_t offset)
+{
+	uint8_t data;
+	switch (offset & 0x0f)
+	{
+		case 0: // RD0
+			data = 0x00;
+			//TODO: if (m_RVD) data |= 0x02;
+			//TODO: if (m_adc->eoc_r()) data |= 0x01;
+			return data;
+		case 1: return ioport("RD1")->read();
+		case 2: return ioport("RD2")->read();
+		case 3: return ioport("RD3")->read();
+		case 4: return ioport("RD4")->read();
+		case 5: return ioport("RD5")->read();
+		case 6: return 0xFF; //TODO: RD6
+		case 7: return 0xFF; //TODO: RD7
+		case 8: return ioport("RD8")->read();
+		case 9: return ioport("RD9")->read();
+		default:
+			return 0xFF;
+	}
+}
+
+uint8_t dfs500_state::cpanel_reg2_r(offs_t offset)
+{
+	switch (offset & 0x07)
+	{
+		case 0: return ioport("RD10")->read();
+		case 1: return ioport("RD11")->read();
+		case 2: return ioport("RD12")->read();
+		default:
+			return 0xFF;
+	}
+}
+
+void dfs500_state::cpanel_reg0_w(offs_t offset, uint8_t data)
+{
+	static const uint8_t ls247_map[16] =
+		{ 0x3f,0x06,0x5b,0x4f,0x66,0x6d,0x7d,0x07,0x7f,0x6f,0x58,0x4c,0x62,0x69,0x78,0x00 };
+
+	switch (offset & 0x0f)
+	{
+		case 0: // WR0 on IC48 KY-223
+			// Dot-points on 7seg digits:
+			// D7 = not connected
+			// D6 = LD206 = DP on trans_rate100
+			// D5 = LD198 = DP on trans_rate10
+			// D4 = LD190 = DP on trans_rate1
+			// D3 = LD182 = DP on pattern_number1000
+			// D2 = LD174 = DP on pattern_number100
+			// D1 = LD186 or LD168 (?) = DP on pattern_number10
+			// D0 = LD158 = DP on pattern_number1
+			m_dot_points2 = data;
+			output().set_value("trans_rate100", ls247_map[m_trans_rate[1] & 0x0F] | (BIT(data >> 6, 0) << 7));
+			output().set_value("trans_rate10", ls247_map[(m_trans_rate[0] >> 4) & 0x0F] | (BIT(data >> 5, 0) << 7));
+			output().set_value("trans_rate1", ls247_map[m_trans_rate[0] & 0x0F] | (BIT(data >> 4, 0) << 7));
+			output().set_value("pattern_number1000", (m_pattern_number[2] & 0x7F) | (BIT(data >> 3, 0) << 7));
+			output().set_value("pattern_number100", (m_pattern_number[1] & 0x7F) | (BIT(data >> 2, 0) << 7));
+			output().set_value("pattern_number10", ls247_map[(m_pattern_number[0] >> 4) & 0x0F] | (BIT(data >> 1, 0) << 7));
+			output().set_value("pattern_number1", ls247_map[m_pattern_number[0] & 0x0F] | (BIT(data >> 0, 0) << 7));
```

Comment:
Tag lookups for outputs are also expensive.

Be consistent with hexadecimal case – you’re mixing uppercase and lowercase.  MAME developers tend to prefer lowercase.

Reply:
Not sure how to do it without these lookups. Is there an example of what you have in mind here?

Reply:
At this separate branch, I've refactored the code to use `output_finder`:
https://github.com/mamedev/mame/commit/aaeb409a399b7d19f225449048bb4c0a7bc222c5

But it seems that initiating output_finder with a list of tags is not supported yet.

I'm looking at `src/emu/devfind.h` as well as at `src/emu/output.h`. It seems that only devfind has support for initializing it with a list of tags. `output_manager::output_finder` does not seem to offer a similar method. It would be very useful though!
(correct me if I am wrong. I only quickly looked at these source files)

Reply:
I’m still not convinced doing the outputs that way is a great idea.

I know this isn’t an issue when there’s only one piece of software on the platform, but having the names in code like this doesn’t scale out when you have more than one thing running on the system.  What the output actually connects to isn’t really the concern of the machine.

For a concrete example of drivers heavily dependent on outputs that work this way, see the Bally pinball drivers:
* The [by17.cpp](https://github.com/mamedev/mame/blob/mame0225/src/mame/drivers/by17.cpp) and [by35.cpp](https://github.com/mamedev/mame/blob/mame0225/src/mame/drivers/by35.cpp) drivers have output names that match the wiring diagrams.
* We have internal artwork for [Mata Hari](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/by17_matahari.lay), [PowerPlay](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/by17_pwerplay.lay) and [Playboy](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/by35_playboy.lay) that shows the playfield elements.
* We also have generic fallback artwork for showing the state of the outputs for the [17](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/by17.lay) and [35](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/by35.lay) boards.

Another example is the BFM Scorpion 2 system, where we have internal artwork for [Cops'n'Robbers Club](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/sc2copcl7.lay), [Premier Club Manager](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/sc2prem2.lay), [Along the Prom](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/sc2prom.lay), [Party Time](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/sc2ptytm1.lay), and some other machines, as well as generic artwork showing the lamp state in a grid, and the DMD and video output if present.

In practical terms, it’s more effort to modify or fix errors in the outputs if you attach semantics to them in the code.  At the very least, you need to go through a compile/link cycle to test a change.  If you’re just sending the raw values to outputs, you can just edit an external layout file to test a change, so the turnaround time for experimentation and testing is a lot faster.  It’s also more effort to keep a list of names in sync with the layouts than just a bunch of positions in the output matrix.

Until recently it wasn’t possible to extract bit fields from an output in the layout code, which would have caused issues for you used a single output for an 8-bit output matrix row.  However, that’s been addressed, and you can extract individual bits or bit fields with `statemask` for conditionally drawn elements, or `mask` for element position/colour animation.

Several internal layouts have been updated to use this functionality already, including:
* https://github.com/mamedev/mame/blob/00bfc6411ad8005edfbc347802d831793da7c177/src/mame/layout/unkeinv.lay#L31 (extracting a bit field from an input for item bounds animation)
* https://github.com/mamedev/mame/blob/00bfc6411ad8005edfbc347802d831793da7c177/src/mame/layout/esq2by16.lay#L35 (extracting individual bits from outputs for alphanumeric dot matrix displays)
* I also posted a simple demo of using `statemask` to allow the use of external graphics for 7-segment displays over at MAMEworld: https://www.mameworld.info/ubbthreads/showthreaded.php?Cat=&Number=388122&page=0&view=expanded&sb=5&o=&fpart=1&vc=1 (it also shows SVG in external artwork, but that’s a different kettle of fish)

But to answer the question about how to get an array of output finders with literal names, if you really need to, you can do it like this using C++ uniform initialisation:

```
class foo : ...
{
...
    output_finder<> m_outputs[10];
};
...
foo::foo(...)
...
    m_outputs{
        { *this, "first_name" },
        { *this, "second_name" },
        ...
        { *this, "tenth_name" } },
...
{
...
}
...
```